gadopt/momentum_equation.py

```diff
@@ -47,195 +42,147 @@ class ViscosityTerm(BaseTerm):
-
-        # NOTE, here source term F is already on the RHS
-        F = dot(phi, source)*self.dx
-
-        return F
-
-
-class MomentumEquation(BaseEquation):
-    """Momentum equation with viscosity, pressure gradient, and source terms."""
-    terms = [ViscosityTerm, PressureGradientTerm, MomentumSourceTerm]
-
-
-class ContinuityEquation(BaseEquation):
-    """Mass continuity equation with a single divergence term."""
-    terms = [DivergenceTerm]
-
-
-def StokesEquations(
-    test_space: fd.functionspaceimpl.WithGeometry,
-    trial_space: fd.functionspaceimpl.WithGeometry,
-    quad_degree: Optional[int] = None,
-    **kwargs,
-) -> list[BaseEquation]:
-    """Stokes system involving the momentum and mass continuity equations.
-
-    Arguments:
-      test_space: Firedrake function space of the test function
-      trial_space: Firedrake function space of the trial function
-      quad_degree: Quadrature degree. Default value is `2p + 1`, where
-                     p is the polynomial degree of the trial space
-
-    Returns:
-      A list of equation instances for the Stokes system.
+    where σ is a penalty parameter.
 
+    Epshteyn, Y., & Rivière, B. (2007).
+    Estimation of penalty parameters for symmetric interior penalty Galerkin methods.
+    Journal of Computational and Applied Mathematics, 206(2), 843-872.
     """
-    mom_eq = MomentumEquation(
-        test_space.sub(0), trial_space.sub(0), quad_degree=quad_degree, **kwargs
-    )
-    cty_eq = ContinuityEquation(
-        test_space.sub(1), trial_space.sub(1), quad_degree=quad_degree, **kwargs
-    )
-    return [mom_eq, cty_eq]
+    dim = eq.mesh.geometric_dimension()
+    identity = Identity(dim)
+    compressible_stress = eq.approximation.compressible
+
+    mu = eq.approximation.mu
+    stress = eq.stress
+    F = inner(nabla_grad(eq.test), stress) * eq.dx
+
+    sigma = interior_penalty_factor(eq)
+    sigma *= FacetArea(eq.mesh) / avg(CellVolume(eq.mesh))
+    if not is_continuous(eq.trial_space):
+        trial_tensor_jump = tensor_jump(eq.n, trial) + tensor_jump(trial, eq.n)
+        if compressible_stress:
+            trial_tensor_jump -= 2 / 3 * identity * jump(trial, eq.n)
+
+        F += (
+            sigma
+            * inner(tensor_jump(eq.n, eq.test), avg(mu) * trial_tensor_jump)
+            * eq.dS
+        )
+        F -= inner(avg(mu * nabla_grad(eq.test)), trial_tensor_jump) * eq.dS
+        F -= inner(tensor_jump(eq.n, eq.test), avg(stress)) * eq.dS
+
+    # NOTE: Unspecified boundaries are equivalent to free stress (i.e. free in all
+    # directions).
+    # NOTE: "un" can be combined with "stress" provided the stress force is tangential
+    # (e.g. no normal flow with wind)
+    for bc_id, bc in eq.bcs.items():
+        if "u" in bc and any(bc_type in bc for bc_type in ["drag", "stress", "un"]):
+            raise ValueError(
+                '"drag", "stress", or "un" cannot be specified if "u" is already given.'
```

Comment:
I think you've removed the "drag" option now

Reply:
Good catch, thank you.